core/engine/engine.go

```diff
@@ -1,99 +1,166 @@
-		c := lnk.(cidlink.Link).Cid
-		val, err := ds.Get(datastore.NewKey(c.String()))
-		if err != nil {
-			return nil, err
-		}
-		return bytes.NewBuffer(val), nil
+func (e *Engine) PublishLocal(ctx context.Context, adv core.Advertisement) (cid.Cid, error) {
+	return e.publishLocal(ctx, adv)
+}
+
+func (e *Engine) publishLocal(ctx context.Context, adv core.Advertisement) (cid.Cid, error) {
+	// Advertisement are published in datastore by the linkSystem when links are generated.
+	// The linksystem of the reference provider determines how to persist the advertisement.
+	adLnk, err := schema.AdvertisementLink(e.lsys, adv)
+	if err != nil {
+		return cid.Undef, err
 	}
-	lsys.StorageWriteOpener = func(_ ipld.LinkContext) (io.Writer, ipld.BlockWriteCommitter, error) {
-		buf := bytes.NewBuffer(nil)
-		return buf, func(lnk ipld.Link) error {
-			c := lnk.(cidlink.Link).Cid
-			return ds.Put(datastore.NewKey(c.String()), buf.Bytes())
-		}, nil
+	lnk, err := adLnk.AsLink()
+	if err != nil {
+		return cid.Undef, err
+	}
+	c := lnk.(cidlink.Link).Cid
+	id, err := adv.ID.AsBytes()
+	if err != nil {
+		return cid.Undef, err
+	}
+	// Store latest advertisement published from the chain.
+	// NOTE: The datastore should be thread-safe, if not
+	// we need a lock to protect races on this value.
+	err = e.putLatestAdv(id)
+	if err != nil {
+		return cid.Undef, err
 	}
-	return lsys
+	return c, nil
 }
 
-func (e *Engine) PublishLocal(ctx context.Context, ad core.Advertisement) error {
-	panic("not implemented")
+func (e *Engine) Publish(ctx context.Context, adv core.Advertisement) error {
+	// First publish the advertisement locally.
+	// Advertisements are stored immutably, so it doesn't matter
+	// if we try to publish the same advertisement twice, we will just
+	// propagate the announcement again through the pubsub channel.
+	c, err := e.publishLocal(ctx, adv)
+	if err != nil {
+		return err
+	}
+
+	// Use legPublisher to publish the advertisement.
+	return e.lp.UpdateRoot(ctx, c)
 }
 
-func (e *Engine) Publish(ctx context.Context, ad core.Advertisement) error {
+func (e *Engine) PushAdv(ctx context.Context, indexer peer.ID, adv core.Advertisement) error {
 	panic("not implemented")
 }
 
-func (e *Engine) PushAdv(ctx context.Context, indexer peer.ID, ad core.Advertisement) error {
+func (e *Engine) Push(ctx context.Context, indexer peer.ID, cid cid.Cid, metadata []byte) {
 	panic("not implemented")
 }
 
-func (e *Engine) Push(ctx context.Context, indexer peer.ID, cid cid.Cid, val indexer.Value) {
-	panic("not implemented")
+func (e *Engine) PutEvent(ctx context.Context, cids []cid.Cid, metadata []byte) error {
+	latestAdvID, err := e.getLatestAdv()
+	if err != nil {
+		return err
+	}
+	latestIndexLink, err := e.getLatestIndexLink()
+	if err != nil {
+		return err
+	}
+	// Lsys will store the index conveniently here.
+	_, indexLnk, err := schema.NewSingleEntryIndex(e.lsys, cids, nil, metadata, latestIndexLink)
```

Comment:
how hard will it be to not store the cids themselves?
we could imagine that we may want to assume we can regenerate them on demand from the car, rather than having them duplicated into the datastore here, since they'll be the bulk of the data

Reply:
Great point. I was deferring this to the future. But I'll look into it right now.